core/src/main/java/io/questdb/griffin/engine/join/NestedLoopLeftJoinRecordCursorFactory.java

```diff
@@ -0,0 +1,173 @@
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.join;
+
+import io.questdb.cairo.AbstractRecordCursorFactory;
+import io.questdb.cairo.sql.*;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.EmptyTableRecordCursor;
+import io.questdb.std.Misc;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Nested Loop with filter join.
+ * Iterates on master factory in outer loop and on slave factory in inner loop
+ * and returns all row pairs matching filter plus all unmatched rows from master factory.
+ **/
+public class NestedLoopLeftJoinRecordCursorFactory extends AbstractRecordCursorFactory {
+    private final NestedLoopLeftRecordCursor cursor;
+    private final Function filter;
+    private final RecordCursorFactory masterFactory;
+    private final RecordCursorFactory slaveFactory;
+
+    public NestedLoopLeftJoinRecordCursorFactory(
+            RecordMetadata metadata,
+            RecordCursorFactory masterFactory,
+            RecordCursorFactory slaveFactory,
+            int columnSplit,
+            @NotNull Function filter,
+            @NotNull Record nullRecord
+    ) {
+        super(metadata);
+        this.masterFactory = masterFactory;
+        this.slaveFactory = slaveFactory;
+        this.filter = filter;
+        this.cursor = new NestedLoopLeftRecordCursor(columnSplit, filter, nullRecord);
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        RecordCursor masterCursor = masterFactory.getCursor(executionContext);
+        RecordCursor slaveCursor = null;
+        try {
+            slaveCursor = slaveFactory.getCursor(executionContext);
+            if (masterCursor.hasNext()) {
+                cursor.of(masterCursor, slaveCursor, executionContext);
+                return cursor;
+            }
+            slaveCursor = Misc.free(slaveCursor);
+            Misc.free(masterCursor);
+        } catch (Throwable ex) {
+            Misc.free(masterCursor);
+            Misc.free(slaveCursor);
+            throw ex;
+        }
+
+        return EmptyTableRecordCursor.INSTANCE;
+    }
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsUpdateRowId(CharSequence tableName) {
+        return masterFactory.supportsUpdateRowId(tableName);
```

Comment:
Not tested. Please test it in `UPDATE` statement.

Reply:
Is there a realistic way to test it ? 
Explicit joins, subqueries and with alias update are not allowed in UPDATE statement. 
This method follows CrossJoin factory . Should I remove it so it returns the default [false] ?

Reply:
You can have joins in updates. This should have a cross join from a to b

```sql
UPDATE a SET r = 1
FROM b
WHERE b.id = 123
```

This will have a left join

```
UPDATE a SET r = 1
FROM b
LEFT JOIN c on b.id = c.id
WHERE a.id = b.id
```

Reply:
That doesn't seem to be the case and only implicit joins via where clause conditions are accepted . 
The update sql you mentioned produces "JOIN is not supported on UPDATE statement" error message .
Please have a look at UpdateTest.testUpdateWithJoinUnsupported .

Reply:
@ideoma anything to add ?

Reply:
@bziobrowski is right, UPDATE with JOINs are prohibited, I created separate issue to update docs or code #2900